state_machine_design.py

```diff
@@ -0,0 +1,355 @@
+    do a bit of processing to make its determination. The example of that
+    would be the SLEEP state which calls into sleep then picks a new state.
+    '''
+
+    def process(self, sm):
+        '''
+        The process function for each state does what the state is meant to accomplish.
+        To be overwritten in each state.
+        '''
+
+    def update(self, sm):
+        '''
+        The update function for each state selects the state to transition to.
+        To be overwritten in each state.
+        '''
+
+
+class WAKE_UP(State):
+    '''
+    The WAKE_UP state is always the first state to run.
+    It makes sure we actually have a file we can run and
+    then passes control to CHECK_TIME.
+    '''
+
+    def process(self, sm):
+        print("WAKE_UP process func")
+
+        # read active ensembles file
+        try:
+            with open(sm.ens_filename, "r", encoding="utf-8") as f_in:
+                sm.ens = json.load(f_in)
+
+            f_in.close()
+            # f = open(sm.ens_filename)
+            # sm.ens = json.load(f)
+            # f.close()
+            sm.ens_index = sm.ens["next_ensemble"]
+            sm.ens_list = sm.ens["ensemble_list"]
+        except:
+            sm.err_code = NO_ENS_FILE
+            sm.err_msg = "active_ensembles.json not found"
+
+        now = time.localtime()
+        curr_time_seconds = hmsToSeconds(now.tm_hour, now.tm_min, now.tm_sec)
+
+        print("waking up at: " + str(curr_time_seconds))
+
+    def update(self, sm):
+        print("WAKE_UP update func")
+        # if active ensembles file not found or other reading with error,
+        # transition to ERROR state
+        if sm.err_code != NO_ERR:
+            sm.state = ERROR()
+        else:
+            # otherwise always
+            # transition to CHECK_TIME state
+            sm.state = CHECK_TIME()
+
+
+class CHECK_TIME(State):
+    '''
+    CHECK_TIME determines if the machine should skip an ensemble,
+    run an ensemble, or sleep until the next ensemble.
+    '''
+
+    def __init__(self):
+        self.check_time_ctrl = SKIP
+
+    def process(self, sm):
+        print("CHECK_TIME process func")
+        print("ens index: " + str(sm.ens_index))
+
+        if sm.ens_index < len(sm.ens_list):
+            # read time from ensemble and compare to current_time
+            now = time.localtime()
+            curr_time_seconds = hmsToSeconds(now.tm_hour, now.tm_min, now.tm_sec)
+
+            nearest_ens_time = sm.ens_list[sm.ens_index]["start_time"]
+
+            if nearest_ens_time < curr_time_seconds:
+                print("time past current ens, skipping")
+                self.check_time_ctrl = SKIP
+                sm.err_code = MISSED_ENS
+                # sm.err_msg = "Current ensemble past time, skipping ensemble "
+                # + sm.ens_list[sm.ens_index]["title"] + " at " + curr_time_seconds
+            elif nearest_ens_time == curr_time_seconds:
```

Comment:
We might want to leave more room for error here. Say, if `nearest_ens_time` hasn't passed yet but is within 5 seconds we could go ahead and execute.

Reply:
Just so I'm understanding correctly, you're saying we can run an ensemble up to about 5 seconds _early_?

Reply:
Yes. We might play with the exact interval, but I think we'd rather execute slightly early than accidentally miss the single second when we should perform the ensemble.